post-processing-code/general-functions/aggregation.py

```diff
@@ -0,0 +1,167 @@
+import glob
+import csv
+import pandas as pd
+import numpy as np
+
+AGGEGATE_DEFAULT_TYPING_MAP = {"year_id": float, "age_start": float, "age_end": float,
+                               "mean": float, "2.5_percentile": float, "5_percentile": float,
+                               "10_percentile": float, "25_percentile": float, "50_percentile": float,
+                               "75_percentile": float, "90_percentile": float, "95_percentile": float,
+                               "97.5_percentile": float, "median": float, "standard_deviation": float
+                               }
+
+def percentile(n):
+        def percentile_(x):
+            return np.percentile(x, n)
+        percentile_.__name__ = '%s_percentile' % n
+        return percentile_
+def calc_not_na(x):
+    return x.notnull().mean() * 100
+
+def aggregatePostProcessedFiles(
+    pathToOutputFiles: str = ".",
+    specific_files: str = "*.csv",
+    columns_to_replace_with_nan: list[str] = ["mean"],
+    typingMap: dict = AGGEGATE_DEFAULT_TYPING_MAP
+) -> pd.DataFrame:
+    """
+    Combines all data outputs in a given folder and filters as necessary. The format of the csv's in the folder should be the same output format of a call to `processSingleFile`
+
+    Args:
+        pathToOutputFiles (str): the top level folder where the output files are located.
+        specific_files (str): file name filter to only combine the files that are wanted.
+        columns_to_replace_with_nan (list[str]): any columns that should be replaced with NaN rather than None.
+        typingMap (dict): a dictionary that maps column names to their dtype.
+    
+    Returns:
+        A dataframe with all of the CSVs stacked ontop of one another
+    """
+    rows = []
+    columns = []
+
+    files_to_combine = glob.glob(pathToOutputFiles + "**/" + specific_files, recursive=True)
+    index = 0
+    total_files = len(files_to_combine)
+    for filename in files_to_combine:
+        index += 1
+        with open(filename, newline="") as f:
+            reader = csv.reader(f)
+            if len(columns) == 0:
+                columns = next(reader)
+            else:
+                next(reader)
+            rows.extend(reader)
+        if ((total_files >= 10) and (index % (total_files // 10) == 0)):
+            print(f"Files Processed: {index / total_files * 100}%")
+    aggregatedData = pd.DataFrame(rows, columns=columns)
+    aggregatedData.loc[:, columns_to_replace_with_nan] = aggregatedData.loc[:, columns_to_replace_with_nan].replace("", np.nan)
+    aggregatedData.loc[:, ~aggregatedData.columns.isin(columns_to_replace_with_nan)] = aggregatedData.loc[:, ~aggregatedData.columns.isin(columns_to_replace_with_nan)].replace("", None)
+    aggregatedData = aggregatedData.astype(typingMap)
+    return aggregatedData
+
+def iuLevelAggregate(
```

Comment:
❓ I don't really understand this function - it seems like in all uses it is a no-op?

Reply:
Yea, I realized as I was coding this that the iuLevelAggregate is basically doing a very similar thing to the general aggregate function. It might make more sense to make this function end up calling the above aggregate function (i.e you pass in a folder with all the files, and it will aggregate and filter everything for you).